module/zfs/arc.c

```diff
@@ -1755,7 +1756,7 @@ arc_hdr_authenticate(arc_buf_hdr_t *hdr, spa_t *spa, uint64_t dsobj)
 		abd_take_ownership_of_buf(abd, B_TRUE);
 
 		csize = zio_compress_data(HDR_GET_COMPRESS(hdr),
-		    hdr->b_l1hdr.b_pabd, tmpbuf, lsize);
```

Comment:
Nope, steps to get problem here:
- Use code from this PR
- get 89+% compressed encrypted block
- L2ARC attached to pool
- Load pool with pre-PR code
- `zio_compress_data` will not compress this block again here due to old 12.5% threshold inside it
  
Heck, because it's the problem - I'll add a read-only feature flag only to make it safe for this case. In future `zio_compress_data` will have threshold argument so we won't get in this probem again.

Reply:
I see, if you move the pool to a system with older software, it doesn't correctly write to the L2ARC, because even though the block was compressed, the old L2ARC-writing code will not compress it if it doesn't save at least 1/8th of the space.  This only applies if the old system has disabled compressed ARC, right?

Introducing a readonly-compatible feature flag is a bit unfortunate.  It would mean maintaining the old 1/8th code forever.  I wonder how common it is to disable compressed ARC, nowadays?  It might be reasonable to say that if you disable compressed ARC, and use L2ARC, and bring a newer pool back to old bits, then some of the data won't be able to "stick" in the L2ARC.  I mean, we previously considered (but didn't implement) saying that disabling compressed ARC makes L2ARC not work with compressed blocks at all (i.e. removing the recompress-on-L2ARC-write code entirely).

Reply:
@ahrens yes, you've got the point. The funny thing, that I can't reproduce problems in such theoretic scenario, l2arc works well under manual tests with data from new code, but with old module and disabled ARC compression. 

There is a "best effort" thing near `arc_hdr_authenticate` function, maybe it's the case?

I've tried to add a feature flag, and code began to transform from beautiful and easy to read info ugly if-if monster. I don't want to do this :(

If it's already a problematic and rare case (compressed l2arc with encryption on old code with pool and data from new code), *and* we are ready to go with this - I propose to merge it as is without a feature flag, PR is ready for it. I can add this comment somewhere (I'd be glad if someone gives me exact place for this).